### PR TITLE
init image at creation time

### LINKED ISSUE #719: Image init data buffer
At the moment clvk defers the initialisation of image content given at creation (either by `CL_MEM_COPY_HOST_PTR` or `CL_MEM_USE_HOST_PTR`).

It leads to a staging buffer being allocated until the data is actually copied to the image memory.

I have an application where lots of images with init data are created and it ends up hitting an out-of-memory.

Could we have a mode where the copy of the data to the image memory is done at creation time, not at first use?

### PR DESCRIPTION
**Pull request by rjodinchr**
fix #719

### CODE REVIEW

File: src/queue.cpp [hunk 1]
```diff
@@ -101,6 +101,10 @@ cl_int cvk_command_queue::satisfy_data_dependencies(cvk_command* cmd) {
             continue;
         }
         CVK_ASSERT(mem->is_image_type());
+        if (config.init_image_at_creation()) {
+            tracker.set_state(cvk_mem_init_state::completed);
```
Comment (kpet) [inline src/queue.cpp:105]: Shouldn't this be done in `cvc_image::init_vulkan_image`, after completing the execution of the init command?

File: src/context.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "context.hpp"
+#include "queue.hpp"
+
+cvk_command_queue* cvk_context::get_or_create_image_init_command_queue() {
```
Comment (kpet) [inline src/context.cpp:18]: Looks like you need some kind of locking or protections. AFACIT after a quick read, it seems two threads could be creating this queue at the same time.

File: src/context.hpp [hunk 3]
```diff
@@ -120,6 +126,8 @@ struct cvk_context : public _cl_context,
     size_t m_printf_buffersize;
     cvk_printf_callback_t m_printf_callback;
     void* m_user_data;
+
+    cvk_command_queue* m_queue_image_init = nullptr;
```
Comment (kpet) [inline src/context.hpp:130]: `std:unique_ptr`? You wouldn't need the free function.

**Review by kpet — CHANGES_REQUESTED** [review]
Thanks! Fine with this being an option. Please add a unit test.

File: src/context.cpp (hunk 2 above)
Comment (rjodinchr) [reply src/context.cpp:18]: I agree, I've added a lock

File: src/queue.cpp (hunk 1 above)
Comment (rjodinchr) [reply src/queue.cpp]: Yes, that's better, changed made

File: src/context.hpp (hunk 3 above)
Comment (rjodinchr) [reply src/context.hpp:135]: I agree, that's what I tried to do initially. But it is not simple. clvk's headers contain too much code everywhere leading to build errors (you can see that `context.hpp` does not include `queue.hpp` in this PR, it's only declaring `struct cvk_command_queue`, thus making the destructor inaccessible for `std::unique_ptr`).

I've tried to move a few things around, but it quickly started to be too much making this PR too complicated for what it is trying to achieve.

I can open an issue to have a rework task for it to do once this land if you want.

Comment (kpet) [reply src/context.hpp:135]: Right, we do need to untangle those headers a bit. I've already bumped into similar issues in the past. Please create an issue with as many specifics as possible. Thanks!

File: src/queue.hpp [hunk 4]
```diff
@@ -175,6 +182,11 @@ struct cvk_command_queue : public _cl_command_queue,
     }
 
     cvk_buffer* get_or_create_printf_buffer() {
+        if (m_init_image_queue) {
+            cvk_error_fn("Could not get printf buffer from init-image queue");
+            CVK_ASSERT(false);
+            return nullptr;
+        }
```
Comment (kpet) [inline src/queue.hpp:189]: Why is this required?

File: src/queue.cpp [hunk 5]
```diff
@@ -63,6 +64,10 @@ cvk_command_queue::cvk_command_queue(
 
     TRACE_CNT(batch_in_flight_counter, 0);
     TRACE_CNT(group_in_flight_counter, 0);
+
+    if (init_image_queue) {
+        m_context.reset(nullptr);
+    }
```
Comment (kpet) [inline src/queue.cpp:70]: Why is this required?

Comment (rjodinchr) [reply src/queue.cpp:70]: `cvk_context` is refcounted. It will not be freed until every object releases their hold on it.
`cvk_command_queue` has a ref on `cvk_context` and that new queue for image initialisation is freed in `cvk_context` destructor with that PR.
Thus to allow everything to get freed properly, the new queue for image initialisation releases the context at creation.

File: src/queue.hpp (hunk 4 above)
Comment (rjodinchr) [reply src/queue.hpp:189]: because the queue for image initialisation does not have a context, which is needed for that function.
Image initialisation should not need printf.

File: src/context.hpp (hunk 3 above)
Comment (rjodinchr) [reply src/context.hpp:135]: https://github.com/kpet/clvk/issues/735

File: src/queue.cpp (hunk 5 above)
Comment (kpet) [reply src/queue.cpp:70]: That makes sense, thanks. Hmm, rather than introducing a new type of queue effectively, with state in every single command queue, we could introduce a `cvk_command_queue::detach_from_context` function doing the above. That function would only be called in `get_or_create_image_init_command_queue` and you could just assert that `m_context` is non-NULL in `get_or_create_printf_buffer` (although this should really never be called if the command queue is only used for image init; part of me thinks this assert could be omitted, the resulting crash would be very obvious and we wouldn't run the risk of adding new functions that use `m_context` without a corresponding assert). WDYT?

Comment (rjodinchr) [reply src/queue.cpp]: I'm fine with that.
Done

**Review by kpet — APPROVED** [review]
Thanks!
